README.md
# **UPDATED FORK**

As the original repository did not receive updates nor PRs for a couple of years, we will use this fork to address some issues and upgrade its capabilities.
Read the changes carefully.

## Addressed issues

* Upgraded dependencies (golang.org/x/crypto) to address Autocert fails to sign certificate automatically with Let's encrypt.
```
acme/autocert: unable to satisfy "https://acme-v02.api.letsencrypt.org/acme/authz-v3/..." for domain "www.mydomain.com": no viable challenge type found
```

## New capabilities

* Added capability to upload via command line utilities. A new button added to the main page UI, where you can get an example cURL command to use for upload:
<p align="center">
  <img alt="uploadCurl" src="media/upload-via-curl.gif" height="500" />
</p>

* When uploading via command line utility you can specify the mime-type via special header **x-pwndrop-content-type**:
```
curl -X POST -H "x-pwndrop-content-type: application/javascript" -H "Authorization: [redacted]" -F "file=@Malicious.exe"  https://<your-server>/api/v1/files
```


## Build from source code

The original installation files were removed as we are going to Go Get the necessary dependencies.
This installation was tested with GO version: **1.18.4**: https://go.dev/dl/

Go Installtion example (https://go.dev/doc/install):
```
rm -rf /usr/local/go && tar -C /usr/local -xzf go1.18.4.linux-amd64.tar.gz
export PATH=$PATH:/usr/local/go/bin
go version #verify at least 1.18
```

Build Pwndrop:
```
git clone https://github.com/Sygnia/pwndrop.git
cd pwndrop
make build
ls build/
```

A 'build' directory will be created under the repository folder.
You can use the compiled binary in the same way described in the original README below.



*****************************
<p align="center">
  <img alt="pwndrop logo" src="https://raw.githubusercontent.com/kgretzky/pwndrop/master/media/pwndrop-logo-512.png" height="120" />
  <p align="center">
    <img alt="pwndrop title" src="https://raw.githubusercontent.com/kgretzky/pwndrop/master/media/pwndrop-title-black-512.png" height="40" />
  </p>
</p>

**pwndrop** is a self-deployable file hosting service for sending out red teaming payloads or securely sharing your private files over HTTP and WebDAV.

If you've ever needed to quickly set up an nginx/apache web server to host your files and you were never happy with the limitations of `python -m SimpleHTTPServer`, **pwndrop** is definitely for you!

<p align="center">
  <img alt="demo" src="https://raw.githubusercontent.com/kgretzky/pwndrop/master/media/demo1.gif" height="500" />
</p>

With **pwndrop** you can:
- [x] Upload and immediately share multiple files using your own private VPS, using drag & drop.
- [x] Decide to make files available or unavailable for download with a single click.
- [x] Set up custom download URLs, for shared files, without playing with directory structure.
- [x] Set up facade files, which will be served instead of the original file whenever you feel like it.
- [x] Set up automatic redirects to spoof the file's extension in a shared link.
- [x] Change MIME type of the served file to change browser's behavior when a download link is clicked.
- [x] Serve files over HTTP, HTTPS and WebDAV.
- [x] Install and setup everything using a bash oneliner.
- [x] Set up **pwndrop** to work as a nameserver and respond with a valid DNS A record to any sub-domain you choose.
- [x] Protect your admin panel behind a custom secret URL path and log in securely with your own username and password.
- [x] Never worry about setting up HTTPS certificates as **pwndrop** does everything for you in the background (including auto-renewals).

Its main goal is to make file sharing as easy and intuitive as possible, while implementing extra features to aid in red team assessments.

Frontend of **pwndrop** is developed in pure Vue.js + Bootstrap with no npm or webpack dependencies. The backend serves REST API and manages a local database, powered by GO language.

## Write-up

If you want to learn how to use **pwndrop** or you want to learn what new features were implemented in recent releases, make sure to check out the posts on my blog:

https://breakdev.org/pwndrop

## Video guide

Take a look at the fantastic video made by Luke Turvey ([@TurvSec](https://twitter.com/TurvSec)), which fully explains how to get started using **pwndrop**.

[![File and Phishing Payload Hosting using PwnDrop (Red Team) - Luke Turvey](https://img.youtube.com/vi/e3veSyIFvOE/0.jpg)](https://www.youtube.com/watch?v=e3veSyIFvOE)

## Prerequisites

If you don't yet have the server to deploy to I highly recommend Digital Ocean. The cheapest $5/mo Debian 9 server with 25GB of storage space will work wonders for you. You can use my referral link to [get an extra $100 to spend on your servers in 60 days for free](https://m.do.co/c/50338abc7ffe).

Register a new domain and point its DNS A records to your VPS IP. You can also register a domain and point its `ns1` and `ns2` nameservers to **pwndrop** instance IP - it will automatically respond with valid DNS A replies.

1. Registered domain name pointing to **pwndrop** instance IP as a DNS A records or as a nameserver.
2. Server with at least 512 MB RAM.

If you want to set up **pwndrop** without a domain, check below how to set up a local instance, which will not auto-generate HTTPS certificates.

## Installation

Make sure there aren't any DNS or HTTP(S) servers running before you attempt to install **pwndrop**.

#### Oneliner

I do not recommend running oneliners, before downloading and checking the script code, but if you are really in a hurry, here it is:
```
curl https://raw.githubusercontent.com/kgretzky/pwndrop/master/install_linux.sh | sudo bash
```

This will download the latest amd64 release binary and fully install a daemon running in a background.

#### From binary

First you need to download the release package you want from: https://github.com/kgretzky/pwndrop/releases

Then do the following (this performs same actions to the oneliner):

```
tar zxvf pwndrop-linux-amd64.tar.gz
./pwndrop stop
./pwndrop install
./pwndrop start
./pwndrop status
```

#### From source code

First of all, make sure you have installed GO with version at least **1.13**: https://golang.org/doc/install

Then do the following:

```
sudo apt-get -y install git make
git clone https://github.com/kgretzky/pwndrop
cd pwndrop
make
make install
```

## Quickstart

Make sure the **pwndrop** is running.

1. Open the secret URL to authorize your browser: `https://yourdomain.com/pwndrop` (this is a default value; make sure to use the secret path, you've pre-configured)
2. Open the admin panel URL in your browser: `https://yourdomain.com/` (since you've authorized your browser, you will now see an admin panel login page)
3. Create your admin account or login.
4. Click the configuration cog in top-left corner and make sure you change the secret path to something other than `/pwndrop`.

You're good to go!

## Running from CLI

You don't have to install **pwndrop** as a daemon and you can run it straight from the console.

```
usage: pwndrop [start|stop|install|remove|status] [-config <config_path>] [-debug] [-no-autocert] [-no-dns] [-h]

daemon management:
    start           : start the daemon
    stop            : stop the daemon
    install         : install the daemon using the available system manager (systemd, systemv and upstart supported)
    remove          : uninstall the daemon
    status          : check status of the installed daemon

parameters:
    -config         : specify a custom path to a config file (def. 'pwndrop.ini' in same directory as the executable)
    -debug          : enable debug output 
    -no-autocert    : disable automatic TLS certificate retrieval from LetsEncrypt; useful when you want to connect over IP or/and in a local network
    -no-dns         : do not run a DNS server on port 53 UDP; use this if you don't want to use pwndrop as a nameserver
    -h              : usage help
```

## Configuration

On first launch, **pwndrop**, by default, will create a new configuration file `pwndrop.ini` in the same directory as an executable. You can later modify it or supply your own, for example to pre-configure **pwndrop** before the installation to automate the deployment of a tool even better.

Here is an example config file with all available config variables with commentary:
```
[pwndrop]
listen_ip = "190.33.86.22"                  # the external IP of your pwndrop instance (must be set if you want to use the nameserver feature)
http_port = 80                              # listening port for HTTP and WebDAV
https_port = 443                            # listening port for HTTPS
data_dir = "./data"                         # directory path where data storage will reside (relative paths are from executable directory path)
admin_dir = "./admin"                       # directory path where the admin panel files reside (relative paths are from executable directory path)

[setup]                                     # optional: put in if you want to pre-configure pwndrop (section will be deleted from the config file on first run)
username = "admin"                          # username of the admin account
password = "secretpassword"                 # password of the admin account
redirect_url = "https://www.somedomain.com" # URL to which visitors will be redirected to if they supply a path, which doesn't point to any shared file (put blank if you want to return 404)
secret_path = "/pwndrop"                    # secret URL path, which upon visiting will allow your browser to access the login page of the admin panel (make sure to change the default value)
```

If you want to pre-configure your **pwndrop** instance before deployment using any of the installation scripts, put your configuration file at `/usr/local/pwndrop/pwndrop.ini` and it will be parsed the moment **pwndrop** daemon is first executed.

## Credits

Huge thanks to [**@jaredhaight**](https://twitter.com/jaredhaight) for inspiring me to learn Vue, with his [Faction C2](https://www.factionc2.com/) framework!

Also much thanks to all the people who gave me pre-release feedback and supported me with their opinions on the tool!

## License

**pwndrop** is made by Kuba Gretzky ([@mrgretzky](https://twitter.com/mrgretzky)) and it's released under GPL3 license.


CHANGELOG.md
****v1.0.3****
- [x] Added capability to upload via command line utilities. This is done by saving a key to the browser's local storage upon a successful login. Users can supply the Authorization header with the key to upload files. Also added a green button to the UI, under 'Upload', so you can copy ready made cURL commands with your key included. Now you can also force the mime-type you would like to set through HTTP header - x-pwndrop-content-type.
- [x] Upgraded included dependencies versions as well as Go version. This fixes the autocert fails issues when requesting the server with a domain.
- [x] Removed unnecessary installation scripts, see read me update section.

****v1.0.2****
- [x] Unauthorized connections not pointing to any hosted files, returning 404, are now automatically closed instead of being kept alive. This resolves the issue of pwndrop getting DDoSed quickly with bots hammering requests at it from various sources.
- [x] Anti-DDoS feature has been added, which temporarily blacklists every IP address of a client who made 10 consecutive requests returning 404. Blacklist period is currently 10 minutes.
- [x] Removed timeouts for uploading and downloading files fully. The previous 15 minutes timeout would have not helped with DDoS attacks anyway. 

****v1.0.1****
- [x] Increased the time limit for uploads and downloads from 15 seconds to 15 minutes. Should fix the issue of uploads/downloads being interrupted on slow connections, when handling big files.

core/server.go
package core

import (
	"crypto/tls"
	"fmt"
	"net"
	"net/http"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/gorilla/mux"
	"golang.org/x/crypto/acme"

	"github.com/kgretzky/pwndrop/api"
	"github.com/kgretzky/pwndrop/log"
	"github.com/kgretzky/pwndrop/storage"
)

const (
	API_PATH = "api/v1"
)

type Server struct {
	srv       *http.Server
	listenTLS net.Listener
	listen    net.Listener
	wdav      *WebDav
	http      *Http
	cdb       *CertDb
	ns        *Nameserver
	r         *mux.Router
	blacklist map[string]*BlacklistItem
	bl_mtx    sync.Mutex
}

func NewServer(host string, port_plain int, port_tls int, enable_letsencrypt bool, enable_dns bool, ch_exit *chan bool) (*Server, error) {
	var err error
	s := &Server{
		blacklist: make(map[string]*BlacklistItem),
		bl_mtx:    sync.Mutex{},
	}

	hostname := fmt.Sprintf("%s:%d", host, port_plain)
	hostname_tls := fmt.Sprintf("%s:%d", host, port_tls)

	s.cdb, err = NewCertDb(Cfg.GetDataDir())
	if err != nil {
		return nil, err
	}

	cert, err := LoadTLSCertificate(filepath.Join(Cfg.GetDataDir(), "public.crt"), filepath.Join(Cfg.GetDataDir(), "private.key"))
	if err != nil {
		log.Warning("certificate: %s", err)
		cert, err = GenerateTLSCertificate(host)
		if err != nil {
			return nil, err
		}
		log.Info("generated self-signed certificate")
	} else {
		log.Info("using TLS certificate from data directory")
		enable_letsencrypt = false
	}

	tls_cfg := &tls.Config{}
	tls_cfg.Certificates = append(tls_cfg.Certificates, *cert)
	if enable_letsencrypt {
		log.Info("autocert: enabled")
		tls_cfg.GetCertificate = s.cdb.AutocertMgr.GetCertificate
		tls_cfg.NextProtos = []string{
			"h2", "http/1.1", // enable HTTP/2
			acme.ALPNProto, // enable tls-alpn ACME challenges
		}
	} else {
		log.Info("autocert: disabled")
	}

	// set up modern cipher suites
	/*
		tls_cfg.MinVersion = tls.VersionTLS12
		tls_cfg.CipherSuites = []uint16{
			tls.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
			tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
			tls.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305, // Go 1.8 only
			tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,   // Go 1.8 only
			tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
			tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,

			// Best disabled, as they don't provide Forward Secrecy,
			// but might be necessary for some clients
			// tls.TLS_RSA_WITH_AES_256_GCM_SHA384,
			// tls.TLS_RSA_WITH_AES_128_GCM_SHA256,
		}*/

	s.wdav, err = NewWebDav(s)
	if err != nil {
		return nil, err
	}
	s.http, err = NewHttp(s)
	if err != nil {
		return nil, err
	}

	s.setupRouter()

	s.srv = &http.Server{
		Handler:      http.Handler(s),
		Addr:         hostname,
		WriteTimeout: 0,
		ReadTimeout:  0,
		IdleTimeout:  5 * time.Second,
		TLSConfig:    tls_cfg,
	}

	s.listenTLS, err = tls.Listen("tcp", hostname_tls, tls_cfg)
	if err != nil {
		return nil, err
	}
	s.listen, err = net.Listen("tcp", hostname)
	if err != nil {
		return nil, err
	}

	log.Info("starting HTTP/WebDAV server at %s", hostname)
	log.Info("starting HTTPS server at %s", hostname_tls)

	if enable_dns {
		s.ns, err = NewNameserver(ch_exit)
		if err != nil {
			return nil, err
		}
	}

	go func() {
		err := s.srv.Serve(s.listen)
		if err != nil {
			log.Fatal("failed to start HTTP/WebDAV server at %s", hostname)
			*ch_exit <- false
		}
	}()

	go func() {
		err := s.srv.Serve(s.listenTLS)
		if err != nil {
			log.Fatal("failed to start HTTPS server at %s", hostname_tls)
			*ch_exit <- false
		}
	}()

	return s, nil
}

func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	log.Debug("%s %s", r.Method, r.URL.Path)

	from_ip := r.RemoteAddr
	if strings.Contains(from_ip, ":") {
		from_ip = strings.Split(from_ip, ":")[0]
	}

	if s.isBlacklisted(from_ip) {
		err := s.killConnection(w, -1)
		if err != nil {
			log.Error("http: %s (%s)", err, from_ip)
			w.Header().Set("Connection", "close")
			w.Header().Set("Content-Length", "0")
			w.WriteHeader(500)
		}
		return
	}

	if !s.isWebDavRequest(r) {

		cookie_name := Cfg.GetCookieName()
		cookie_token := Cfg.GetCookieToken()

		if r.URL.Path == Cfg.GetSecretPath() {
			ck := &http.Cookie{
				Domain:   "",
				Path:     "/",
				Expires:  time.Now().AddDate(0, 3, 0),
				HttpOnly: true,
				Name:     cookie_name,
				Value:    cookie_token,
			}
			http.SetCookie(w, ck)
			http.Redirect(w, r, "/", http.StatusFound)
			return
		}

		if !s.FileExists(r.URL.Path) {
			if ck, err := r.Cookie(cookie_name); err == nil {
				if ck.Value == cookie_token {
					s.r.ServeHTTP(w, r)
					return
				}
			}

			if _, err := api.AuthApiKey(r); err == nil {
				s.r.ServeHTTP(w, r)
				return
			}
			
			s.addBlacklistHit(from_ip)
			if len(Cfg.GetRedirectUrl()) > 0 {
				http.Redirect(w, r, Cfg.GetRedirectUrl(), http.StatusFound)
				return
			}
		}

		// http
		s.http.ServeHTTP(w, r)
	} else {
		// webdav
		s.wdav.Handler().ServeHTTP(w, r)
	}
}

func (s *Server) isWebDavRequest(r *http.Request) bool {
	ua := r.Header.Get("user-agent")
	if strings.Index(ua, "WebDAV") >= 0 || strings.Index(ua, "DavClnt") >= 0 {
		return true
	}
	if r.Header.Get("translate") == "f" {
		return true
	}
	return false
}

func (s *Server) setupRouter() {
	admin_path := "/"
	s.r = mux.NewRouter()
	sr := s.r.PathPrefix(admin_path + API_PATH).Subrouter()
	sr.HandleFunc("/auth", api.AuthOptionsHandler).Methods("OPTIONS")
	sr.HandleFunc("/auth", api.AuthCheckHandler).Methods("GET")
	sr.HandleFunc("/server_info", api.ServerInfoOptionsHandler).Methods("OPTIONS")
	sr.HandleFunc("/server_info", api.ServerInfoGetHandler).Methods("GET")
	sr.HandleFunc("/version", api.VersionOptionsHandler).Methods("OPTIONS")
	sr.HandleFunc("/version", api.VersionGetHandler).Methods("GET")
	sr.HandleFunc("/login", api.AuthOptionsHandler).Methods("OPTIONS")
	sr.HandleFunc("/login", api.LoginUserHandler).Methods("POST")
	sr.HandleFunc("/logout", api.LogoutUserHandler).Methods("GET")
	sr.HandleFunc("/clear_secret", api.ClearSecretSessionHandler).Methods("GET")
	sr.HandleFunc("/create_account", api.AuthOptionsHandler).Methods("OPTIONS")
	sr.HandleFunc("/create_account", api.CreateUserHandler).Methods("POST")
	sr.HandleFunc("/config", api.ConfigOptionsHandler).Methods("OPTIONS")
	sr.HandleFunc("/config", api.ConfigGetHandler).Methods("GET")
	sr.HandleFunc("/config", api.ConfigUpdateHandler).Methods("POST")
	sr.HandleFunc("/files", api.FileOptionsHandler).Methods("OPTIONS")
	sr.HandleFunc("/files", api.FileListHandler).Methods("GET")
	sr.HandleFunc("/files", api.FileCreateHandler).Methods("POST")
	sr.HandleFunc("/files/{id}", api.FileDeleteHandler).Methods("DELETE")
	sr.HandleFunc("/files/{id}", api.FileUpdateHandler).Methods("PUT")
	sr.HandleFunc("/files/{id}/sub", api.SubFileCreateHandler).Methods("POST")
	sr.HandleFunc("/files/{id}/sub/{sub_id}", api.SubFileDeleteHandler).Methods("DELETE")
	sr.HandleFunc("/files/{id}/enable", api.FileEnableHandler).Methods("GET")
	sr.HandleFunc("/files/{id}/disable", api.FileDisableHandler).Methods("GET")
	sr.HandleFunc("/files/{id}/pause", api.FilePauseHandler).Methods("GET")
	sr.HandleFunc("/files/{id}/unpause", api.FileUnpauseHandler).Methods("GET")
	s.r.PathPrefix(fmt.Sprintf("%s", admin_path)).Handler(http.StripPrefix(fmt.Sprintf("%s", admin_path), http.FileServer(http.Dir(Cfg.GetAdminDir()))))
}

func (s *Server) handleNotFound(w http.ResponseWriter, r *http.Request) {
	log.Debug("%s %s", r.Method, r.URL.Path)
}

func (s *Server) GetFile(url string) (*storage.DbFile, int, error) {
	is_redirect := false
	f, err := storage.FileGetByUrl(url)
	if err != nil {
		f, err = storage.FileGetByRedirectUrl(url)
		if err != nil {
			return nil, 404, err
		}
		is_redirect = true
	}
	if !f.IsEnabled {
		return nil, 404, fmt.Errorf("file is disabled")
	}
	if f.IsPaused {
		if f.RedirectPath != "" && is_redirect {
			return nil, 404, fmt.Errorf("can't access facade via redirect while paused")
		} else if f.RefSubFile > 0 {
			sf, err := storage.SubFileGet(f.RefSubFile)
			if err != nil {
				return nil, 404, fmt.Errorf("facade file not found")
			}
			f.Filename = sf.Filename
			f.FileSize = sf.FileSize
		} else {
			return nil, 404, fmt.Errorf("facade file not set")
		}
	}
	return f, 200, nil
}

func (s *Server) FileExists(url string) bool {
	_, err := storage.FileGetByUrl(url)
	if err != nil {
		_, err = storage.FileGetByRedirectUrl(url)
		if err != nil {
			return false
		}
	}
	return true
}

func (s *Server) killConnection(w http.ResponseWriter, status int) error {
	if status > 0 {
		w.Header().Set("Connection", "close")
		w.Header().Set("Content-Length", "0")
		w.WriteHeader(status)
	}

	hj, ok := w.(http.Hijacker)
	if !ok {
		return fmt.Errorf("connection hijacking not supported")
	}
	conn, _, err := hj.Hijack()
	if err != nil {
		return err
	}
	conn.Close()
	return nil
}

func (s *Server) isBlacklisted(ip_addr string) bool {
	s.bl_mtx.Lock()
	defer s.bl_mtx.Unlock()

	ret := false
	if bl, ok := s.blacklist[ip_addr]; ok {
		if bl.hits >= BLACKLIST_HITS_LIMIT {
			if time.Now().Before(bl.last_hit.Add(BLACKLIST_JAIL_TIME_SECS * time.Second)) {
				ret = true
			} else {
				delete(s.blacklist, ip_addr)
				return false
			}
		}
		bl.last_hit = time.Now()
	}
	return ret
}

func (s *Server) addBlacklistHit(ip_addr string) {
	s.bl_mtx.Lock()
	defer s.bl_mtx.Unlock()

	if bl, ok := s.blacklist[ip_addr]; ok {
		bl.hits += 1
	} else {
		bl := &BlacklistItem{
			hits:     1,
			last_hit: time.Now(),
		}
		s.blacklist[ip_addr] = bl
	}
}
